SwiftCompilerSources/Sources/SIL/ApplySite.swift

```diff
@@ -91,41 +112,57 @@ extension ApplySite {
     SubstitutionMap(bridged.ApplySite_getSubstitutionMap())
   }
 
+  /// Get the conventions of the callee without the applied substitutions.
+  public var originalCalleeConvention: FunctionConvention {
+    FunctionConvention(for: callee.type.bridged.getASTType(), in: parentFunction)
+  }
+
+  /// Get the conventions of the callee with the applied substitutions.
+  public var substitutedCalleeConvention: FunctionConvention {
+    FunctionConvention(for: bridged.ApplySite_getSubstitutedCalleeType(), in: parentFunction)
+  }
+
+  public var calleeArgumentConventions: ArgumentConventions {
+    ArgumentConventions(functionConvention: substitutedCalleeConvention)
+  }
+
+  public var operandConventions: ApplyOperandConventions {
+    ApplyOperandConventions(
+      calleeArgumentConventions: calleeArgumentConventions,
+      unappliedArgumentCount: bridged.PartialApply_getCalleeArgIndexOfFirstAppliedArg())
+  }
+
   /// Returns the argument index of an operand.
   ///
   /// Returns nil if 'operand' is not an argument operand. This is the case if
   /// it's the callee function operand.
-  public func argumentIndex(of operand: Operand) -> Int? {
-    let opIdx = operand.index
-    if opIdx >= ApplyOperands.firstArgumentIndex &&
-       opIdx <= ApplyOperands.firstArgumentIndex + numArguments {
-      return opIdx - ApplyOperands.firstArgumentIndex
-    }
-    return nil
+  public func calleeArgumentIndex(of operand: Operand) -> Int? {
+    operandConventions.calleeArgumentIndex(of: operand)
   }
 
   /// Returns true if `operand` is the callee function operand and not am argument operand.
-  public func isCalleeOperand(_ operand: Operand) -> Bool {
-    return operand.index < ApplyOperands.firstArgumentIndex
+  public func isCallee(operand: Operand) -> Bool {
+    operandConventions.isCallee(operand: operand)
   }
-  
-  public func getArgumentConvention(calleeArgIndex: Int) -> ArgumentConvention {
-    return bridged.ApplySite_getArgumentConvention(calleeArgIndex).convention
+
+  public func convention(of operand: Operand) -> ArgumentConvention? {
+    operandConventions.convention(of: operand)
+  }
+
+  // Converts an argument index of a callee to the corresponding apply operand.
```

Comment:
Can you move the comment of the original `callerArgIndex(calleeArgIndex: Int)` here (with `s/caller indices/caller operands/`)? Something like
```
  /// If the apply does not actually apply that argument, it returns nil.
  ///
  /// Example:
  /// ```
  ///                 func callee(v, w, x, y, z) { }
  /// // caller operands in %pa:  -, -, c, d, e     ("-" == nil)
  /// // caller operands in %a:   a, b, -, -, -
  ///
  /// %pa = partial_apply @callee(c, d, e)
  ///  %a = apply         %pa    (a, b)
  /// ```
```

Reply:
Yes, that's great!